packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -92,44 +93,74 @@ export class ModularChangeFamily
+                              ...fieldChange,
+                              revision: change.revision,
+                              isInverse: change.isInverse,
+                          };
+
+                getOrAddEmptyToMap(fieldChanges, key).push(fieldChangeToCompose);
             }
         }
 
         const composedFields: FieldChangeMap = new Map();
         for (const [field, changesForField] of fieldChanges) {
-            const { fieldKind, changesets } = this.normalizeFieldChanges(changesForField);
-            const composedField = fieldKind.changeHandler.rebaser.compose(
-                changesets,
-                this.childComposer,
-            );
+            let composedField: FieldChange;
+            if (changesForField.length === 1) {
+                composedField = changesForField[0];
+            } else {
+                const { fieldKind, changesets } = this.normalizeFieldChanges(changesForField);
+                assert(
+                    changesets.length === changesForField.length,
+                    "Number of changes should be constant when normalizing",
+                );
+                const taggedChangesets: TaggedChange<FieldChangeset>[] = [];
+                for (let i = 0; i < changesets.length; i++) {
+                    const taggedChange: TaggedChange<FieldChangeset> = {
+                        revision: changesForField[i].revision,
+                        isInverse: changesForField[i].isInverse,
+                        change: changesets[i],
+                    };
+
+                    taggedChangesets.push(taggedChange);
+                }
+
+                const composedChange = fieldKind.changeHandler.rebaser.compose(
+                    taggedChangesets,
+                    this.childComposer,
+                );
+
+                composedField = {
+                    fieldKind: fieldKind.identifier,
+                    change: brand(composedChange),
+                };
+            }
 
             // TODO: Could optimize by checking that composedField is non-empty
-            composedFields.set(field, {
-                fieldKind: fieldKind.identifier,
-                change: brand(composedField),
-            });
+            composedFields.set(field, composedField);
         }
         return composedFields;
     }
 
-    private composeNodeChanges(changes: NodeChangeset[]): NodeChangeset {
-        const fieldChanges = [];
+    private composeNodeChanges(changes: TaggedChange<NodeChangeset>[]): NodeChangeset {
+        const fieldChanges: TaggedChange<FieldChangeMap>[] = [];
         let valueChange: ValueChange | undefined;
         for (const change of changes) {
-            if (change.valueChange !== undefined) {
-                valueChange = change.valueChange;
+            if (change.change.valueChange !== undefined) {
+                valueChange = change.change.valueChange;
+                valueChange.revision ??= change.revision;
```

Comment:
Don't we want the revision of the last value change here? (even if it's undefined).
So I would have expeted `valueChange.revision = change.revision`.

Reply:
We've just reassigned `valueChange`. If it already had a revision (because the changeset it is part of is a composite changeset), we don't want to overwrite it. However, we don't want to mutate the input changesets, so I suppose we should be cloning `valueChange` before mutating it.

Reply:
Ah sorry, that makes sense to me now. I agree we ought to clone.